File: coursework_console/OrderComponents/DataStorage.cpp
#include "DataStorage.h"

void DataStorage::operator=(DataStorage* other) {
	this->capacity = other->capacity;
	this->transferInterface = other->transferInterface;
	this->brand = other->brand;
	this->formFactor = other->formFactor;
}

DataStorage::DataStorage(DataTransferInterface transferInterface) {
	this->transferInterface = transferInterface;
}

DataStorage::DataStorage(int capacity, DataTransferInterface transferInterface, std::string brand, float formFactor) {
	tryToSetArguments(capacity, transferInterface, brand, formFactor);
}

int DataStorage::getCapacity() const {
	return capacity;
}

DataTransferInterface DataStorage::getInterface() const {
	return transferInterface;
}

std::string DataStorage::getBrand() const {
	return brand;
}

float DataStorage::getFormFactor() const {
	return formFactor;
}

void DataStorage::input() {
	int capacity;
	DataTransferInterface interface;
	std::string brand;
	float formFactor;

	std::cout << "Введите вместимость (в ГБ): ";
	std::cin >> capacity;
	std::cout << "Введите интерфейс подключения (PATA - 0, SATA - 1, SAS - 2, NVMe - 3): ";
	std::cin >> interface;
	std::cin.clear();
	while (std::cin.get() != '\n');
	std::cout << "Введите производителя: ";
	std::getline(std::cin, brand);
	std::cout << "Введите форм фактор (в дюймах): ";
	std::cin >> formFactor;
	std::cin.clear();
	while (std::cin.get() != '\n');

	tryToSetArguments(capacity, interface, brand, formFactor);
}


std::string DataStorage::toString() const {
	std::string name = this->getBrand() + ", " + std::to_string(this->getCapacity()) + " ГБ, " + dataTransferInterfaceToString(this->getInterface()) + ", " + std::format("{:.1f}", this->getFormFactor());
	return name;
}

json DataStorage::toJSON() const {
    json j;
	j["componentName"] = getComponentName();
	j["capacity"] = capacity;
	j["transferInterface"] = dataTransferInterfaceToString(transferInterface);
	j["brand"] = brand;
	j["formFactor"] = formFactor;
	return j;
}

void DataStorage::fromJSON(json data) {
	tryToSetArguments(data["capacity"], stringToDataTransferInterface(data["transferInterface"]), data["brand"], data["formFactor"]);
}


bool DataStorage::checkArguments(int capacity, DataTransferInterface transferInterface, std::string brand, float formFactor) const {
	return capacity >= 0 && formFactor >= 0;
}

void DataStorage::tryToSetArguments(int capacity, DataTransferInterface transferInterface, std::string brand, float formFactor) {
	if (checkArguments(capacity, transferInterface, brand, formFactor)) {
		this->capacity = capacity;
		this->transferInterface = transferInterface;
		this->brand = brand;
		this->formFactor = formFactor;
	}
	else throw std::invalid_argument("Некорректный формат данных!");
}

File: courseworkWinFormsCSharp/OrderComponents/DataStorage.cs
using Newtonsoft.Json.Linq;
using System;

namespace courseworkWinFormsCSharp.OrderComponents
{
    public abstract class DataStorage : ILaptopComponent
    {
        private int _capacity;
        public int Capacity
        {
            get => _capacity;
            set
            {
                if (value >= 0)
                    _capacity = value;
                else
                    throw new ArgumentException("Некорректный объем памяти!");
            }
        }
        public DataTransferInterface TransferInterface { get; set; } = DataTransferInterface.SATA;
        public string Brand { get; set; } = string.Empty;

        private float _formFactor;
        public float FormFactor
        {
            get => _formFactor;
            set
            {
                if (value >= 0)
                    _formFactor = value;
                else
                    throw new ArgumentException("Некорректный форм-фактор!");
            }
        }

        public DataStorage() { }

        public DataStorage(DataTransferInterface TransferInterface)
        {
            this.TransferInterface = TransferInterface;
        }

        public DataStorage(int Capacity, DataTransferInterface TransferInterface, string Brand, float FormFactor)
        {
            this.Capacity = Capacity;
            this.TransferInterface = TransferInterface;
            this.Brand = Brand;
            this.FormFactor = FormFactor;
        }

        public DataStorage(JObject json)
        {
            FromJSON(json);
        }

        public abstract string GetComponentName();


        public override string ToString()
        {
            return $"{Brand}, {Capacity} ГБ, {TransferInterface}, {FormFactor:F1}";
        }

        public virtual JObject ToJSON()
        {
            JObject json = new JObject
            {
                { "componentName", GetComponentName() },
                { "capacity", Capacity },
                { "transferInterface", TransferInterface.ToString() },
                { "brand", Brand },
                { "formFactor", FormFactor }
            };
            return json;
        }

        public virtual void FromJSON(JObject data)
        {
            Capacity = data.GetValue("capacity").ToObject<int>();
            TransferInterface = DataTransferInterfaceConverter.StringToDataTransferInterface(data.GetValue("transferInterface").ToString());
            Brand = data.GetValue("brand").ToString();
            FormFactor = data.GetValue("formFactor").ToObject<float>();
        }
    }
}
